### PR TITLE
[Fix] RawAudio.toBlob() ignores TypedArray byteOffset/length (#1704)

### LINKED ISSUE #1704: RawAudio.toBlob() encodes the entire underlying ArrayBuffer, ignoring TypedArray byteOffset/length
### System Info

  `@huggingface/transformers` 4.2.0 (web, also reproduces in Node). Regression from v3.x — 3.5.1 encodes the same input correctly.

  ### Description
  
  `encodeWAV` builds the output blob from each sample array's whole backing buffer:

  ```js
  return new Blob([buffer, ...chunks.map((chunk) => chunk.buffer)], { type: "audio/wav" });
  ```

  The WAV header's data-chunk size is computed from the **view's** length, but `chunk.buffer` contributes the **entire** underlying `ArrayBuffer`. When the audio
  is a `subarray()` view (byteOffset ≠ 0, or length < buffer length), header and payload disagree. Decoders honor the header, so they play the first N samples of
  the buffer — the wrong window of audio.
  
  ### Reproduction

  ```js
  import { RawAudio } from "@huggingface/transformers";

  const full = new Float32Array(48000);                                  // 2s @ 24kHz
  for (let i = 24000; i < 48000; i++) full[i] = Math.sin(i / 20) * 0.5;  // tone in 2nd half

  const view = full.subarray(24000);                                     // keep only the tone
  const blob = new RawAudio(view, 24000).toBlob();

  console.log(blob.size);
  // actual:   192044  (44-byte header + the whole 2s buffer)
  // expected:  96044  (44-byte header + the 1s view)
  ```

  Decoding the blob (e.g. `decodeAudioData`) plays the first second — silence — instead of the tone.
  
  ### Real-world impact

  In a streaming TTS pipeline that trims leading silence with `subarray()` before calling `toBlob()`, every start-trim silently becomes an end-truncation: the
  chunk keeps its leading silence and loses real speech off the end.

  ### Suggested fix

  Respect the view's window when building the blob:
  
  ```js
  return new Blob(
    [buffer, ...chunks.map((c) => new Uint8Array(c.buffer, c.byteOffset, c.byteLength))],
    { type: "audio/wav" }
  );
  ```
  
  (Passing the typed-array views directly also works — `Blob` parts honor `ArrayBufferView` offset/length.)

### PR DESCRIPTION
**Pull request by yushuosun**
## Motivation

`RawAudio.toBlob()` produces a WAV whose audio is wrong (e.g. leading silence / only the first N samples) when the audio is a `TypedArray` view with a non-zero `byteOffset` or a length shorter than its backing buffer — anything coming from `Float32Array.subarray(...)`, for example (#1704).

## Root cause

In `packages/transformers/src/utils/audio.js`, `encodeWAV()` writes the WAV data-chunk size from each view's length (`chunk.length`), but builds the `Blob` payload from `chunk.buffer` — the entire underlying `ArrayBuffer`, ignoring the view's `byteOffset`/`byteLength`. Header and payload then disagree, and decoders honor the header, reading the wrong window of samples.

## Modifications

`packages/transformers/src/utils/audio.js` (`encodeWAV`): build the payload from `Uint8Array` views that respect each chunk's `byteOffset`/`byteLength`, instead of passing the whole `chunk.buffer`.

## Duplicate-check

- Track A — `gh api 'repos/huggingface/transformers.js/pulls?state=open&per_page=100' --paginate` grepped for `1704` → no open PR references it.
- Track B — issue #1704 has no comments and no in-progress claim.

### CODE REVIEW

File: packages/transformers/src/utils/audio.js [hunk 1]
```diff
@@ -813,7 +813,7 @@ function encodeWAV(chunks, rate) {
     /* data chunk length */
     view.setUint32(40, totalLength * 4, true);
 
-    return new Blob([buffer, ...chunks.map((chunk) => /** @type {ArrayBuffer} */ (chunk.buffer))], {
+    return new Blob([buffer, ...chunks.map((chunk) => new Uint8Array(chunk.buffer, chunk.byteOffset, chunk.byteLength))], {
```
Comment (nico-martin) [inline packages/transformers/src/utils/audio.js]: The fix is correct. Please add a regression test using a non-zero-offset `Float32Array.subarray()`, asserting the Blob size (`44 + view.byteLength`), WAV data length, and that payload samples come from the selected view rather than the start of its backing buffer.

**Review by nico-martin — CHANGES_REQUESTED** [review]
Hi @yushuosun, thank you so much for looking into this! The fix is correct and nicely minimal. Could you add a small regression test using a non-zero-offset `Float32Array.subarray()`? I'm holding off on merging until that test protects the exact issue from #1704. Since #1714 is the same fix, I'd keep this earlier PR.

**Comment by yushuosun** [comment]
Thanks for the review, @nico-martin! Added a regression test in 32b49f4 using a non-zero-offset `Float32Array.subarray()` view — it asserts the Blob size (`44 + view.byteLength`), the WAV `data` chunk length, and that payload samples come from the selected view rather than the start of the backing buffer.

Verified it fails on the previous implementation (`Expected: 60, Received: 72`, i.e. the whole backing buffer leaking into the WAV) and passes with the fix. And sounds good — happy for this PR to be the canonical one for the fix.

**Review by nico-martin — CHANGES_REQUESTED** [review]
Thanks for adding the focused regression test. It covers exactly the requested non-zero-offset case and passes locally. There is one build issue left at head `32b49f4`: `tsc --build` rejects the new `Uint8Array` as `Uint8Array<ArrayBufferLike>` because `BlobPart` requires an `ArrayBuffer`-backed view. A cast would hide the possible `SharedArrayBuffer` case, so please keep the zero-copy view when `chunk.buffer instanceof ArrayBuffer` and use `.slice()` to copy only the shared-backed case. I validated that approach against the 22 focused tests, TypeScript, and a shared-buffer subarray. `audio.js` also currently fails Prettier, so please format it in the same follow-up. After those two mechanical fixes, this is ready to approve.

**Comment by yushuosun** [comment]
Done in df73e2a — `encodeWAV` now keeps the zero-copy `Uint8Array` view only when `chunk.buffer instanceof ArrayBuffer`, and falls back to `chunk.slice()` for shared-backed chunks (no cast, so the `SharedArrayBuffer` case stays explicit rather than hidden). Also ran Prettier over `audio.js` in the same commit.

Locally verified: `tsc --build` passes, all 22 audio tests pass, and `prettier --check` is clean. Thanks for catching the `Uint8Array<ArrayBufferLike>` vs `BlobPart` nuance!

**Review by nico-martin — APPROVED** [review]
Thanks for the follow-up! LGTM!

File: packages/transformers/src/utils/audio.js [hunk 2]
```diff
@@ -813,9 +813,18 @@ function encodeWAV(chunks, rate) {
...
-        type: 'audio/wav',
-    });
+    // SharedArrayBuffer cannot back a BlobPart, so only ArrayBuffer-backed chunks get a zero-copy view.
+    return new Blob(
+        [
+            buffer,
+            ...chunks.map((chunk) =>
+                chunk.buffer instanceof ArrayBuffer
+                    ? new Uint8Array(chunk.buffer, chunk.byteOffset, chunk.byteLength)
+                    : chunk.slice(),
```
[the remark this answers is not in the document]
Comment (xenova) [reply packages/transformers/src/utils/audio.js:823]: good point. @yushuosun would you mind added a test case for `SharedArrayBuffer` too? 👼 thanks!

[the remark this answers is not in the document]
Comment (xenova) [reply packages/transformers/src/utils/audio.js:823]: can easily be done in a follow-up... no need to block this PR

**Review by xenova — APPROVED** [review]
thanks!

**Comment by xiankaing** [comment]
Thanks for the fix!
